### ISSUE TITLE
Refactor syncing code to use requests by hashes rather than by numbers

### ISSUE THREAD

**Issue by arkpar** [ethcore/parity#696]
Split the chain into subchains of 65536, download them sequentially
For each subchain: split it into chunks of 256 blocks by hash, download them in parallel.

Have 3 states: 1. Finding common block -> 2. Downloading subchain chunk headers -> 3. Downloading the rest of the subchain headers and bodies -> 1.

**Comment by arkpar** [comment]
Implemented in master
